## MLP for Binary Classification
https://machinelearningmastery.com/tensorflow-tutorial-deep-learning-with-tf-keras
This dataset involves predicting whether a structure is in the atmosphere or not, given radar returns.

We will use a LabelEncoder to encode the string labels to integer values 0 and 1. The model will be fit on 67% of the data, and the remaining 33% will be used for evaluation, split using the train_test_split() function.

It is good practice to use ‘relu‘ activation with a ‘he_normal‘ weight initialization. This combination goes a long way in overcoming the problem of vanishing gradients when training deep neural network models. For more on ReLU, see the tutorial:

The model predicts the probability of class 1 and uses the sigmoid activation function. The model is optimized using the adam version of stochastic gradient descent and seeks to minimize the cross-entropy loss.

In [1]:
import tensorflow as tf
import pandas as pd

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

In [3]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

In [4]:
import warnings
warnings.filterwarnings('ignore')

In [5]:
tf.get_logger().setLevel('ERROR')

## Load the Dataset

In [6]:
url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/ionosphere.csv'
df = pd.read_csv(url, header=None)

In [7]:
df

0   1        2        3        4        5        6        7        8   \
0     1   0  0.99539 -0.05889  0.85243  0.02306  0.83398 -0.37708  1.00000   
1     1   0  1.00000 -0.18829  0.93035 -0.36156 -0.10868 -0.93597  1.00000   
2     1   0  1.00000 -0.03365  1.00000  0.00485  1.00000 -0.12062  0.88965   
3     1   0  1.00000 -0.45161  1.00000  1.00000  0.71216 -1.00000  0.00000   
4     1   0  1.00000 -0.02401  0.94140  0.06531  0.92106 -0.23255  0.77152   
..   ..  ..      ...      ...      ...      ...      ...      ...      ...   
346   1   0  0.83508  0.08298  0.73739 -0.14706  0.84349 -0.05567  0.90441   
347   1   0  0.95113  0.00419  0.95183 -0.02723  0.93438 -0.01920  0.94590   
348   1   0  0.94701 -0.00034  0.93207 -0.03227  0.95177 -0.03431  0.95584   
349   1   0  0.90608 -0.01657  0.98122 -0.01989  0.95691 -0.03646  0.85746   
350   1   0  0.84710  0.13533  0.73638 -0.06151  0.87873  0.08260  0.88928   

          9   ...       25       26       27       28       29       30  \
0    0.03760  ... -0.51171  0.41078 -0.46168  0.21266 -0.34090  0.42267   
1   -0.04549  ... -0.26569 -0.20468 -0.18401 -0.19040 -0.11593 -0.16626   
2    0.01198  ... -0.40220  0.58984 -0.22145  0.43100 -0.17365  0.60436   
3    0.00000  ...  0.90695  0.51613  1.00000  1.00000 -0.20099  0.25682   
4   -0.16399  ... -0.65158  0.13290 -0.53206  0.02431 -0.62197 -0.05707   
..       ...  ...      ...      ...      ...      ...      ...      ...   
346 -0.04622  ... -0.04202  0.83479  0.00123  1.00000  0.12815  0.86660   
347  0.01606  ...  0.01361  0.93522  0.04925  0.93159  0.08168  0.94066   
348  0.02446  ...  0.03193  0.92489  0.02542  0.92120  0.02242  0.92459   
349  0.00110  ... -0.02099  0.89147 -0.07760  0.82983 -0.17238  0.96022   
350 -0.09139  ... -0.15114  0.81147 -0.04822  0.78207 -0.00703  0.75747   

          31       32       33  34  
0   -0.54487  0.18641 -0.45300   g  
1   -0.06288 -0.13738 -0.02447   b  
2   -0.24180  0.56045 -0.38238   g  
3    1.00000 -0.32382  1.00000   b  
4   -0.59573 -0.04608 -0.65697   g  
..       ...      ...      ...  ..  
346 -0.10714  0.90546 -0.04307   g  
347 -0.00035  0.91483  0.04712   g  
348  0.00442  0.92697 -0.00577   g  
349 -0.03757  0.87403 -0.16243   g  
350 -0.06678  0.85764 -0.06151   g  

[351 rows x 35 columns]

In [8]:
# split into input and output columns
X, y = df.values[:, :-1], df.values[:, -1]

In [9]:
# ensure all data are floating point values
X = X.astype('float32')

In [10]:
# encode strings to integer
y = LabelEncoder().fit_transform(y)

In [11]:
# split into train and test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(235, 34) (116, 34) (235,) (116,)


In [12]:
# determine the number of input features
n_features = X_train.shape[1]

## Step 1: Define the Model

In [32]:
# define model
model = Sequential()
model.add(Dense(10, activation='relu', kernel_initializer='he_normal', input_shape=(n_features,)))
model.add(Dense(8, activation='relu', kernel_initializer='he_normal'))
model.add(Dense(1, activation='sigmoid'))

## Step 2: Compile the Model

In [39]:
# compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

## Step 3: Fit the Model

In [42]:
# fit the model
model.fit(X_train, y_train, epochs=150, batch_size=32, verbose=0)

## Step 4: Evaluate the Model

In [43]:
# Evaluate the model
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print('Test Accuracy: %.3f' % acc)

Test Accuracy: 0.853


## Step 5: Make Prediction

In [44]:
# make a prediction
row = [1,0,0.99539,-0.05889,0.85243,0.02306,0.83398,-0.37708,1,0.03760,0.85243,-0.17755,0.59755,-0.44945,0.60536,-0.38223,0.84356,-0.38542,0.58212,-0.32192,0.56971,-0.29674,0.36946,-0.47357,0.56811,-0.51171,0.41078,-0.46168,0.21266,-0.34090,0.42267,-0.54487,0.18641,-0.45300]
y_pred = model.predict([row])
print('Predicted: %.3f' % y_pred)

1/1 [==============================] - 0s 83ms/step
Predicted: 1.000


In [45]:
y_pred

array([[0.9998553]], dtype=float32)

The model achieved a classification accuracy of about 96% and then predicted a probability of 0.999 that the one row of data belongs to class 1.